### PR TITLE
Implement APIs for parsing expressions, types, paths, patterns and items

### PR DESCRIPTION
**Pull request by davidlattimore**

### CODE REVIEW

**Comment by davidlattimore** [comment]
Just discovered that parsing `a + b` as a path "succeeded", resulting in the path `a`. I added a test for this and fixed by checking that the TokenSource as at EOF after parsing finished.

File: crates/ra_syntax/src/lib.rs [hunk 1]
```diff
@@ -168,6 +168,41 @@ impl SourceFile {
     }
 }
 
+impl crate::ast::Path {
```
Comment (matklad) [inline crates/ra_syntax/src/lib.rs:171]: Let's use `impl ast::Path` here and elsewhere. In retrospect, `pub use crate::ast::SourceFile;` was a bad idea, but that should be fixed in a separate PR. 

Otherwise, LGTM!

bors d+

**Comment by matklad** [comment]
@davidlattimore note that you don't need to wait for CI status to become green before issuing `bors r+`: bors runs exactly the same set of checks. The PR status is there only to give for folks who can't r+ themselves.
